pkg/gds/admin.go

```diff
@@ -145,6 +147,123 @@ func (s *Admin) setupRoutes() (err error) {
 	return nil
 }
 
+// Summary provides aggregate statistics that describe the state of the GDS.
+func (s *Admin) Summary(c *gin.Context) {
+	// Prepare the output response
+	out := &admin.SummaryReply{
+		Statuses: make(map[string]int),
+		CertReqs: make(map[string]int),
+	}
+
+	// Query the list of VASPs from the data store to perform aggregation counts.
+	iter := s.db.ListVASPs()
+	for iter.Next() {
+		// Fetch VASP from the database
+		vasp := iter.VASP()
```

Comment:
You're right that it could be nil and you're also right that we need to handle this case here. With the `iter.Next()` pattern, you're usually expecting `Next()` to be false if there is an error performing the iteration - this requires the VASP to be preloaded into the iterator on the call to `Next()` and simply returned on `iter.VASP()`. However, in this case - to save computation and memory, the call to iter.VASP() is what loads the VASP into memory and could potentially return nil. I suppose I need to decide if following the pattern so we don't make these kind of mistakes in the future is the right thing to do or if the optimization is worth it and performing the nil check here. Any thoughts?

Good catch!

Reply:
Parsing the `[]byte` data from the database is a potentially heavy (and error-prone, as you pointed out) operation that might not be necessary - e.g. if you're just iterating over keys, counting the number of records, or if you're just grabbing the Honu metadata. If the parsing happens in `iter.Next()` it will always happen no matter if you're just doing the above, which also means a parsing error will halt the iteration even for a key scan. 

I wonder if we should make this `vasp, err := iter.VASP()` to separate the parsing error from the iteration error. It requires more error handling by the user but is more flexible when determining what is going wrong in the database. 

Great questions!

Reply:
Thanks, those use cases where we don't want to necessarily read in the data of every record do seem relevant to a directory service. I think your idea of returning an error along with the data is great, it would definitely help avoid missing handling the error cases in the future.